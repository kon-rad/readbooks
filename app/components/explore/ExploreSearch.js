// @flow

import React from 'react';
import './ExploreSearch.css';
import Autocomplete from '../Autocomplete';
import type { SearchResults } from '../../types/BooksAPI';

type Props = {
  handleSearch: (query: string) => void
};
type State = {
  query: string,
  value: string,
  currentIndex: number,
  open: boolean
};

class ExploreSearch extends React.Component<Props, State> {
  suggestions = [];

  onKeyDown = (event: KeyboardEvent): void => {
    const currentIndex = this.state.currentIndex;
    switch (event.key) {
      case 'ArrowDown':
        if (currentIndex < this.suggestions.length - 1)
          this.setState({ currentIndex: currentIndex + 1 });
        if (!this.state.open) this.setState({ open: true });
        event.preventDefault();
        break;
      case 'ArrowUp':
        if (currentIndex > -1)
          this.setState({ currentIndex: currentIndex - 1 });
        else this.setState({ open: false });
        event.preventDefault();
        break;
      case 'Escape':
        this.setState({ open: false, currentIndex: -1 });
        event.preventDefault();
        break;
      case 'Enter':
        if (this.suggestions.length > 0 && currentIndex >= 0) {
          this.setState({
            value: this.suggestions[currentIndex].volumeInfo.title,
            open: false,
            currentIndex: -1
          });
          event.preventDefault();
        } else {
          this.performSearch(null);
        }
        break;
      default:
        return;
    }
  };

  onSuggestionClick = (hit: string): void => {
    this.setState({ value: hit, open: false, currentIndex: -1 });
    this.performSearch();
  };

  onSuggestions = (hits: Array<SearchResults>): void => {
    this.suggestions = hits;
    if (!this.state.open) this.setState({ open: true });
    this.setState({ currentIndex: -1 });
  };

  updateQuery = (event: SyntheticInputEvent<HTMLInputElement>): void => {
    return this.setState({
      query: event.target.value,
      value: event.target.value
    });
  };

  performSearch = (
    event: ?SyntheticInputEvent<HTMLInputElement>
  ): void => {
    this.setState({ open: false });
    this.props.handleSearch(this.state.query);
    if (event) event.preventDefault();
  };

  constructor(props: Props) {
    super(props);

    this.state = {
      query: '',
      value: '',
      open: false,
      currentIndex: -1
    };
  }

  render() {
    return (
      <div className="search_container">
        <h3>Find your next favorite book!</h3>
        <div className="titleSearchWrapper">
          <label htmlFor="titleSearch" className="titleSearchLabel">
            Search:
          </label>
          <div className="autoCompWrapper">
            <input
              id="titleSearch"
              value={this.state.value}
              onKeyDown={this.onKeyDown}
              onChange={this.updateQuery}
              className="searchInput"
              type="search"
              autoComplete="off"
            />
            <Autocomplete
              currentIndex={this.state.currentIndex}
              query={this.state.query}
              onClick={this.onSuggestionClick}
              onSuggestions={this.onSuggestions}
              open={this.state.open}
            />
          </div>
          <button
            onClick={this.performSearch}
            type="submit"
            className="titleSearchSubmit"
            >
              <i className="circular search link icon"></i>
          </button>
        </div>
      </div>
    );
  }
}

export default ExploreSearch;
