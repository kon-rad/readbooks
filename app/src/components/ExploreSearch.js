import React from 'react';


class ExploreSearch extends React.Component {

  render() {
    return (
      <div>
        <h3>Find interesting reads</h3>
        <input type="text"/>
      </div>
    )
  }
}

export default ExploreSearch;