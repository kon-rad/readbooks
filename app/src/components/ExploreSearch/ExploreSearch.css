
.search_input {
    margin: 0 auto;
    width: 300px;
    border-radius: 4px;
    border: solid 1px grey;

    padding: 10px 20px;
}

.search_container {
    width: 100%;
    display: flex;
    flex-direction: column;
    align-items: center;
}

.autocomplete {
    width: 300px;
    height: auto;
    border: solid 1px grey;
}

.autoCompContainer {
    position: relative;
    width:300px;
    margin: 0 auto;
}
.autoCompList {
    background-color: #ffffff;
    margin: 0;
    padding: 0;
    list-style: none;
    width: 300px;
    overflow: hidden;
    height: auto;
    border: solid 1px grey;
    position: absolute;
    top: 0;
}

.autoCompListItem {
    padding: 4px 8px;
    height: 24px;
    overflow: hidden;
    line-height: 16px;
}

.autoCompListItem:hover {
    background-color: lightcyan;
    cursor: pointer;
}