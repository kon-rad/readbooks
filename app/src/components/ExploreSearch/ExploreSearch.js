// @flow

import React from 'react';
import style from './ExploreSearch.css';

type Props = {};

class ExploreSearch extends React.Component<Props, {keyword: ?string}> {

  constructor(props: Props) {
    super(props);
    // (this: any).handleChange = this.handleChange.bind(this);

    this.state = {
      keyword: ''
    };
  }

  handleChange = (event: SyntheticInputEvent<HTMLInputElement>) => {
    this.setState({ keyword: event.target.value });
  };


  render() {
    return (
      <div className={ style.search_container }>
        <h3>Find interesting reads</h3>
        <input onChange={this.handleChange} className={ style.search_input } type="text"/>
      </div>
    )
  }
}

export default ExploreSearch;