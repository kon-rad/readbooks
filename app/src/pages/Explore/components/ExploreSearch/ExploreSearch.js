// @flow

import React from 'react';
import styles from './ExploreSearch.css';
import Autocomplete from '../Autocomplete/Autocomplete';
import type { TitleResults } from '../../../../types/BooksAPI';

type Props = {};
type State = {
  query: string,
  value: string,
  currentIndex: number,
  open: boolean
}

class ExploreSearch extends React.Component<Props, State> {

  suggestions = [];

  onKeyDown = (event: KeyboardEvent) => {
    let currentIndex = this.state.currentIndex;
    switch(event.key) {
      case 'ArrowDown':
        if (currentIndex < this.suggestions.length - 1) this.setState({currentIndex: currentIndex + 1});
        if (!this.state.open) this.setState({open: true});
        event.preventDefault();
        break;
      case 'ArrowUp':
        if (currentIndex > -1) this.setState({currentIndex: currentIndex - 1});
        else this.setState({open:false});
        event.preventDefault();
        break;
      case 'Escape':
        this.setState({open: false, currentIndex: -1 });
        event.preventDefault();
        break;
      case 'Enter':
        if (this.suggestions.length > 0 && currentIndex >= 0) {
          this.setState({value: this.suggestions[currentIndex].volumeInfo.title, open: false, currentIndex: -1 });
          event.preventDefault();
        }
        break;
      default:
        return;
    }
  };

  onSuggestionClick = (hit: string) => {
    this.setState({ value: hit, open: false, currentIndex: -1 });
  };

  onSuggestions = (hits: Array<TitleResults>) => {
    this.suggestions = hits;
    if (!this.state.open) this.setState({open: true});
    this.setState({ currentIndex: -1 });
  };

  updateQuery = (event: SyntheticInputEvent<HTMLInputElement>) => {
    return this.setState({ query: event.target.value, value: event.target.value });
  };

  constructor(props: Props) {
    super(props);

    this.state = {
      query: '',
      value: '',
      open: false,
      currentIndex: -1
    };
  }

  render() {
    return (
      <div className={styles.search_container}>
        <h3>Find interesting reads</h3>
        <div className={styles.titleSearchWrapper}>
          <label htmlFor="titleSearch" className={styles.titleSearchLabel}>Search By Title:</label>
          <div className={styles.autoCompWrapper}>
            <input
              id="titleSearch"
              value={this.state.value}
              onKeyDown={this.onKeyDown}
              onChange={this.updateQuery}
              className={styles.searchInput}
              type="search"
              autoComplete="off"
            />
            <Autocomplete
              currentIndex={this.state.currentIndex}
              query={this.state.query}
              onClick={this.onSuggestionClick}
              onSuggestions={this.onSuggestions}
              open={this.state.open}
            />
          </div>
          <input type="submit" value="submit" className={styles.titleSearchSubmit}/>
        </div>
      </div>
    )
  }
}

export default ExploreSearch;