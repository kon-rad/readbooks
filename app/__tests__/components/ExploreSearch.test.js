import React from 'react';
import ExploreSearch from '../../components/ExploreSearch';
import { shallow } from 'enzyme';

describe('ExploreSearch', () => {

  it('renders correctly', () => {
    const wrapper = shallow(<ExploreSearch/>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should initialize with state value', () => {
    const wrapper = shallow(<ExploreSearch/>);
    wrapper.find('input[type="search"]').simulate('change', { target: { value: 'some query' }});
    expect(wrapper.state('query')).toEqual('some query');
    expect(wrapper.state('value')).toEqual('some query');
  });

  describe('should respond to input from keyboard', () => {
    let wrapper, event;

    beforeEach(() => {
      wrapper = shallow(<ExploreSearch/>);
      wrapper.instance().suggestions = [
        {
          id: 'resultID1',
          volumeInfo: {
            title: 'Title 1'
          }
        },
        {
          id: 'resultID2',
          volumeInfo: {
            title: 'Title 2'
          }
        }
      ];
      event = { preventDefault: jest.fn() };
    });

    it('should increment currentIndex count on ArrowDown', () => {
      wrapper.find('input[type="search"]').simulate('keydown', { key: 'ArrowDown', ...event });
      expect(wrapper.state('currentIndex')).toEqual(0);
      expect(wrapper.state('open')).toBe(true);
      expect(event.preventDefault).toHaveBeenCalled();
    });

    it('should not increment currentIndex on ArrowDown when at end of list', () => {
      wrapper.setState({ currentIndex: 1 });
      wrapper.find('input[type="search"]').simulate('keydown', { key: 'ArrowDown', ...event });
      expect(wrapper.state('currentIndex')).toEqual(1);
      expect(wrapper.state('open')).toBe(true);
      expect(event.preventDefault).toHaveBeenCalled();
    });

    it('should decrement currentIndex on ArrowUp', () => {
      wrapper.setState({ currentIndex: 1, open: true });
      wrapper.find('input[type="search"]').simulate('keydown', { key: 'ArrowUp', ...event });
      expect(wrapper.state('currentIndex')).toEqual(0);
      expect(wrapper.state('open')).toBe(true);
      expect(event.preventDefault).toHaveBeenCalled();
    });

    it('should decrement currentIndex on ArrowUp and close when at first item', () => {
      wrapper.setState({ currentIndex: -1, open: true });
      wrapper.find('input[type="search"]').simulate('keydown', { key: 'ArrowUp', ...event });
      expect(wrapper.state('currentIndex')).toEqual(-1);
      expect(wrapper.state('open')).toBe(false);
      expect(event.preventDefault).toHaveBeenCalled();
    });

    it('should not decrement currentIndex on ArrowUp when at beginning', () => {
      wrapper.setState({ currentIndex: -1 });
      wrapper.find('input[type="search"]').simulate('keydown', { key: 'ArrowUp', ...event });
      expect(wrapper.state('currentIndex')).toEqual(-1);
      expect(wrapper.state('open')).toBe(false);
      expect(event.preventDefault).toHaveBeenCalled();
    });

    it('should close list on Escape', () => {
      wrapper.setState({ currentIndex: 1, open: true });
      wrapper.find('input[type="search"]').simulate('keydown', { key: 'Escape', ...event });
      expect(wrapper.state('currentIndex')).toEqual(-1);
      expect(wrapper.state('open')).toBe(false);
      expect(event.preventDefault).toHaveBeenCalled();
    });

    it('should set value, reset currentIndex and close list on Enter', () => {
      wrapper.setState({ currentIndex: 1, open: true });
      wrapper.find('input[type="search"]').simulate('keydown', { key: 'Enter', ...event });
      expect(wrapper.state('currentIndex')).toEqual(-1);
      expect(wrapper.state('open')).toBe(false);
      expect(wrapper.state('value')).toEqual('Title 2');
      expect(event.preventDefault).toHaveBeenCalled();
    });
  });
});
