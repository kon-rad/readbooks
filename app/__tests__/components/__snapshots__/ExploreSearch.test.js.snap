// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExploreSearch renders correctly 1`] = `
<div
  className="search_container"
>
  <h3>
    Find interesting reads
  </h3>
  <div
    className="titleSearchWrapper"
  >
    <label
      className="titleSearchLabel"
      htmlFor="titleSearch"
    >
      Search By Title:
    </label>
    <div
      className="autoCompWrapper"
    >
      <input
        autoComplete="off"
        className="searchInput"
        id="titleSearch"
        onChange={[Function]}
        onKeyDown={[Function]}
        type="search"
        value=""
      />
      <Autocomplete
        currentIndex={-1}
        onClick={[Function]}
        onSuggestions={[Function]}
        open={false}
        query=""
      />
    </div>
    <input
      className="titleSearchSubmit"
      type="submit"
      value="submit"
    />
  </div>
</div>
`;
